### PR TITLE
implemented tauID sf

### PR DESCRIPTION
**Pull request by valeriadamante**
This PR is made to implement the tau ID SF correction

### CODE REVIEW

File: tau.py [hunk 1]
```diff
@@ -30,3 +35,18 @@ def getES(self, df, source_dict):
...
 
         return df, source_dict
+
+    def getSF(self, df, source_dict, deepTauVersion):
+        for source in [ central ] + TauCorrProducer.SFSources_dm+ TauCorrProducer.SFSources_pt:
+            updateSourceDict(source_dict, source, 'Tau')
+            genuineTau_SFtype = 'dm' if source in TauCorrProducer.SFSources_dm else 'pt'
+            for scale in getScales(source):
+                syst_name = getSystName(source, scale)
+                df = df.Define(f'Tau_p4_{syst_name}', f'''::correction::TauCorrProvider::getGlobal().getSF(
```
Comment (kandrosov) [inline tau.py:45]: - WPs and deepTau version should be defined in the config and WPs depend on the channel
- apply only to the HTT legs
- store both pt and dm SFs
- keep separate lists for each source type

File: tau.h [hunk 2]
```diff
@@ -85,6 +115,34 @@ class TauCorrProvider : public CorrectionsBase<TauCorrProvider> {
...
 
+    float getSF(const LorentzVectorM& Tau_p4, int Tau_decayMode, int Tau_genMatch, const std::string& wpVSe, const std::string& wpVSmu, const std::string& wpVSjet, UncSource source, UncScale scale, const std::string& genuineTau_SFtype) const
+    {
+        if(isTwoProngDM(Tau_decayMode)) throw std::runtime_error("no SF for two prong tau decay modes");
+        const GenLeptonMatch genMatch = static_cast<GenLeptonMatch>(Tau_genMatch);
+        if(genMatch == GenLeptonMatch::Tau) {
+            const UncScale tau_had_scale = sourceApplies(source, Tau_p4, Tau_decayMode, genMatch)
+                                           ? scale : UncScale::Central;
+            const std::string& scale_str = getScaleStr(tau_had_scale);
+            if(wpVSe < 3 || wpVSmu < 4){
```
Comment (kandrosov) [inline tau.h:126]: - define WP enums instead directly using 3&4
- only enter inside if when tau_had_scale != central
- evaluate for non-central is common -> avoid code duplications
- for high pt taus additional unc is bigger than 5%

File: tau.py [hunk 3]
```diff
@@ -30,3 +35,18 @@ def getES(self, df, source_dict):
...
+                               Tau_p4_{nano}, Tau_decayMode, Tau_genMatch, {TauCorrProducer.deepTauVersion}VSe,
+                               {TauCorrProducer.deepTauVersion}VSmu, {TauCorrProducer.deepTauVersion}VSjet,
+                               ::correction::TauCorrProvider::UncSource::{source}, ::correction::UncScale::{scale},
+                               {genuineTau_SFtype})''')
+                df = df.Define(f'Tau_p4_{syst_name}_delta', f'Tau_p4_{syst_name} - Tau_p4_{nano}')
```
Comment (kandrosov) [inline tau.py:50]: it is not needed to define delta

**Review by kandrosov — CHANGES_REQUESTED** [review]
missing SF corresponding code in corrections.py and in anaTupleProducer

File: tau.py (hunk 3 above)
Comment (valeriadamante) [reply tau.py:50]: ok

File: tau.h (hunk 2 above)
Comment (valeriadamante) [reply tau.h:126]: ok

File: Corrections.py [hunk 4]
```diff
@@ -95,20 +113,37 @@ def getNormalisationCorrections(df, config, sample):
...
-    return df, [ scale ]
+    all_sources = set()
+    all_branches = []
+    df,tau_SF_branches = tau.getSF(df, config)
+    all_sources.update(tau_SF_branches.keys())
+    all_branches.append(tau_SF_branches)
+    all_sources.update(pu_SF_branches.keys())
+    all_branches.append(pu_SF_branches)
+    all_weights = []
+    all_sources.remove(central)
```
Comment (kandrosov) [inline Corrections.py:136]:
```suggestion
    df, tau_SF_branches = tau.getSF(df, config)
    all_branches = [ pu_SF_branches, tau_SF_branches ] 
    all_sources = set(itertools.chain.from_iterable(all_branches))
    all_sources.remove(central)
    all_weights = []
```

File: CorrectionsCore.py [hunk 5]
```diff
@@ -25,5 +25,11 @@ def updateSourceDict(source_dict, source, obj):
...
+        #pass
         raise RuntimeError(f"addUncSource: dupblicated {source} definition for {obj}")
-    source_dict[source].append(obj)
\ No newline at end of file
+    source_dict[source].append(obj)
+
+def CreateSourceDict(self, source_dict, obj, sources):
+        for source in  sources:
+            updateSourceDict(source_dict, source, obj)
+        return source_dict
```
Comment (kandrosov) [inline CorrectionsCore.py:35]:
```suggestion
```

File: Corrections.py [hunk 6]
```diff
@@ -95,20 +113,37 @@ def getNormalisationCorrections(df, config, sample):
...
+    for syst_name in [central] + list(all_sources):
+        branches = getBranches(syst_name, all_branches)
+        product = ' * '.join(branches)
+        weight_name = f'weight_{syst_name}'
+        if syst_name != central:
+            weight_name+='_rel'
+        df = df.Define(f'weight_{syst_name}', f'static_cast<float>(genWeightD * {lumi} * {stitching_weight_string} * {product})')
+        df = df.Define(f'weight_{syst_name}_rel', f'static_cast<float>(weight_{syst_name}/weight_{central})')
+        all_weights.append(weight_name)
```
Comment (kandrosov) [inline Corrections.py:145]:
```suggestion
        weight_name = f'weight_{syst_name}'
        weight_rel_name = weight_name + '_rel'
        weight_out_name = weight_name if syst_name == central else weight_rel_name
        df = df.Define(weight_name, f'static_cast<float>(genWeightD * {lumi} * {stitching_weight_string} * {product})')
        df = df.Define(weight_rel_name, f'static_cast<float>(weight_{syst_name}/weight_{central})')
        all_weights.append(weight_out_name)
```

File: tau.h [hunk 7]
```diff
@@ -85,10 +112,48 @@ class TauCorrProvider : public CorrectionsBase<TauCorrProvider> {
...
+        const GenLeptonMatch genMatch = static_cast<GenLeptonMatch>(Tau_genMatch);
+        if(genMatch == GenLeptonMatch::Tau) {
+            const UncScale tau_had_scale = sourceApplies(source, Tau_p4, Tau_decayMode, genMatch)
+                                           ? scale : UncScale::Central;
+            const std::string& scale_str = getScaleStr(tau_had_scale);
+            const auto sf = tau_vs_jet_->evaluate({Tau_p4.pt(),Tau_decayMode, Tau_genMatch, wpVSjet.first, scale_str, genuineTau_SFtype});
+            if(tau_had_scale != UncScale::Central && (wpVSe.second < static_cast<int>(WorkingPointsTauVSe::VLoose) || wpVSmu.second < static_cast<int>(WorkingPointsTauVSmu::Tight))){
+                const auto sf_central = tau_vs_jet_->evaluate({Tau_p4.pt(), Tau_decayMode, Tau_genMatch,  wpVSjet.first, getScaleStr(UncScale::Central), genuineTau_SFtype});
+                float additional_unc = Tau_p4.pt()<100 ? 0.05 : 0.15;
+                return sf_central * ( ( sf / sf_central )+ additional_unc );
```
Comment (kandrosov) [inline tau.h:131]: this doesn't work if sf < sf_central

File: tau.py [hunk 8]
```diff
@@ -2,25 +2,96 @@
...
+class WorkingPointsBoostedTauVSjet:
+   VVLoose = 1
+   VLoose = 2
+   Loose = 3
+   Medium = 4
+   Tight = 5
+   VTight = 6
+   VVTight = 7
+
+globals()['WorkingPointsTauVSmu']
```
Comment (kandrosov) [inline tau.py:40]:
```suggestion
```
use definition from c++, when needed

File: pu.py [hunk 9]
```diff
@@ -25,9 +25,14 @@ def __init__(self, period):
             puWeightProducer.initialized = True
 
     def getWeight(self, df):
```
Comment (kandrosov) [inline pu.py:27]: add `return_variations` (same logic as for tau)

**Review by kandrosov — CHANGES_REQUESTED** [review]

File: Corrections.py [hunk 10]
```diff
@@ -64,7 +70,12 @@ def applyScaleUncertainties(df):
                 if obj not in source_objs:
                     suffix = 'Central' if obj in [ "Tau", "MET" ] else 'nano'
                     df = df.Define(f'{obj}_p4_{syst_name}', f'{obj}_p4_{suffix}')
-    return df, syst_dict
+    return df,syst_dict
+
+def applySFUncertainties(df):
+    if not initialized:
+        raise RuntimeError('Corrections are not initialized')
+    return tau.getSF(df)
```
[the remark this answers is not in the document]
Comment (valeriadamante) [reply Corrections.py:78]: ok

File: Corrections.py (hunk 4 above)
Comment (valeriadamante) [reply Corrections.py:136]: ok

File: Corrections.py (hunk 6 above)
Comment (valeriadamante) [reply Corrections.py:145]: ok

File: CorrectionsCore.py (hunk 5 above)
Comment (valeriadamante) [reply CorrectionsCore.py:35]: ok

File: tau.h (hunk 7 above)
Comment (valeriadamante) [reply tau.h:131]: ok

File: Corrections.py [hunk 11]
```diff
@@ -43,10 +43,16 @@ def Initialize(period):
...
     pu = puWeightProducer(period=period_names[period])
-    tau = TauCorrProducer(period=period_names[period])
+    tau = TauCorrProducer(period_names[period], config)
     met = METCorrProducer()
     initialized = True
+    syst_dict = { }
+    source_dict = { }
+
+def createSourceDict(sources, obj):
+    updateSourceDict(self.source_dict, sources, obj)
```
[the remark this answers is not in the document]
Comment (valeriadamante) [reply Corrections.py:55]: ok

File: pu.py (hunk 9 above)
Comment (valeriadamante) [reply pu.py:27]: ok

File: tau.py (hunk 8 above)
Comment (valeriadamante) [reply tau.py]: ok

File: Corrections.py [hunk 12]
```diff
@@ -1,13 +1,17 @@
 import os
 import ROOT
 import yaml
+import itertools
 from RunKit.sh_tools import sh_call
+from Common.Utilities import *
```
Comment (kandrosov) [inline Corrections.py:6]: why do we need `Common.Utilities` here?

File: Corrections.py [hunk 13]
```diff
@@ -1,13 +1,17 @@
...
 
 from .tau import TauCorrProducer
 from .met import METCorrProducer
 from .pu import puWeightProducer
 from .CorrectionsCore import *
 
+for wpcl in [WorkingPointsTauVSe,WorkingPointsTauVSmu,WorkingPointsTauVSjet]:
+    ROOT.gInterpreter.Declare(f'{generate_enum_class(wpcl)}')
```
Comment (kandrosov) [inline Corrections.py:14]: this code should be in  baseline Initialize, no in corrections

File: CorrectionsCore.py [hunk 14]
```diff
@@ -25,5 +23,7 @@ def updateSourceDict(source_dict, source, obj):
     if source not in source_dict:
         source_dict[source] = []
     if obj in source_dict[source]:
+        #pass
```
Comment (kandrosov) [inline CorrectionsCore.py:26]: remove

File: tau.h [hunk 15]
```diff
@@ -85,10 +112,49 @@ class TauCorrProvider : public CorrectionsBase<TauCorrProvider> {
...
+        if(genMatch == GenLeptonMatch::Tau) {
+            const UncScale tau_had_scale = sourceApplies(source, Tau_p4, Tau_decayMode, genMatch)
+                                           ? scale : UncScale::Central;
+            const std::string& scale_str = getScaleStr(tau_had_scale);
+            const auto sf = tau_vs_jet_->evaluate({Tau_p4.pt(),Tau_decayMode, Tau_genMatch, wpVSjet.first, scale_str, genuineTau_SFtype});
+            if(tau_had_scale != UncScale::Central && (wpVSe.second < static_cast<int>(WorkingPointsTauVSe::VLoose) || wpVSmu.second < static_cast<int>(WorkingPointsTauVSmu::Tight))){
+                const auto sf_central = tau_vs_jet_->evaluate({Tau_p4.pt(), Tau_decayMode, Tau_genMatch,  wpVSjet.first, getScaleStr(UncScale::Central), genuineTau_SFtype});
+                float additional_unc = Tau_p4.pt()<100 ? 0.05 : 0.15;
+                float final_sf = sf > sf_central ? sf_central * ( ( sf / sf_central )+ additional_unc ) : sf_central * ( ( sf / sf_central )- additional_unc );
+                return final_sf;
```
Comment (kandrosov) [inline tau.h:132]:
```suggestion
                const float additional_unc = Tau_p4.pt() > 100 ? 0.15 : 0.05;
                return sf_central * ( sf / sf_central + std::copysign(additional_unc, sf - sf_central));
```

File: tau.py [hunk 16]
```diff
@@ -30,3 +66,41 @@ def getES(self, df, source_dict):
...
+    #                syst_name = getSystName(source, scale)
+    #                SF_branches[syst_name]= []
+    #                for leg_idx in [0,1]:
+    #                    df = df.Define(f"tau_{leg_idx+1}_DM", f'''httCand.leg_type[{leg_idx}] == Leg::tau ? Tau_decayMode.at(httCand.leg_index[{leg_idx}]) : -1.f;''')
+    #                    df = df.Define(f"tau{leg_idx+1}_TrgSF_{syst_name}",
+    #                                f'''::correction::TauCorrProvider::getGlobal().getTrgSF(
+    #                               httCand.leg_p4[{leg_idx}].Pt(), tau_{leg_idx+1}_DM, trg_type, httCand.channel(), \"sf\"
+    #                               ::correction::TauCorrProvider::UncSource::{source}, ::correction::UncScale::{scale}) ''')
+    #                    SF_branches[syst_name].append(f"tau{leg_idx+1}_idSF_{syst_name}")
+    #        return df,SF_branches
```
Comment (kandrosov) [inline tau.py:106]: remove from this PR

**Review by kandrosov — CHANGES_REQUESTED** [review]

File: Corrections.py (hunk 12 above)
Comment (valeriadamante) [reply Corrections.py:6]: bc in utilities the ```generate_enum_class``` is defined

File: Corrections.py (hunk 13 above)
Comment (valeriadamante) [reply Corrections.py:14]: ok

File: Corrections.py (hunk 12 above)
Comment (valeriadamante) [reply Corrections.py:6]: ok I applied the following comment so no more needed to import utilities

File: CorrectionsCore.py (hunk 14 above)
Comment (valeriadamante) [reply CorrectionsCore.py:26]: ok

File: tau.h (hunk 15 above)
Comment (valeriadamante) [reply tau.h:132]: ok

**Review by kandrosov — APPROVED** [review]

**Pull request merged**